lib/spack/spack/package_prefs.py

```diff
@@ -159,14 +159,21 @@ def has_preferred_targets(cls, pkg_name):
     @classmethod
     def preferred_variants(cls, pkg_name):
         """Return a VariantMap of preferred variants/values for a spec."""
+        variants = []
+        # We want to collect the variants of both the package itself and the
+        # 'all' variants to merge them.
         for pkg in (pkg_name, 'all'):
-            variants = get_packages_config().get(pkg, {}).get('variants', '')
-            if variants:
-                break
-
-        # allow variants to be list or string
-        if not isinstance(variants, string_types):
-            variants = " ".join(variants)
+            v = get_packages_config().get(pkg, {}).get('variants', '')
+            if v:
+                # Allow variants to be list or string
+                if not isinstance(v, string_types):
+                    v = " ".join(v)
+                variants.append(v)
```

Comment:
This seems too early to turn the list of variants into string. I think we should ensure they are treated as a list here, then deduplicated, then turned into strings. Does it make sense?

Reply:
Variant lists are strings.  But semantically, they are sets.  Proper variant merging requires that the strings be parsed into sets ASAP, and then appropriate merge semantics applied to them.  They can then be converted back to strings, if one wishes.

> is supposed to behave, which as far as I understand is the motivation behind your fixme?

See #14448.  In an ideal world, we would have a `link_type` multi-valued variant with values `shared|static`.  The above would then be:
```
packages:
  all:
    variants: ['link_type=static']
  opencv:
    variants: ['link_type=shared']
```
In this case, the `link_type` for `opencv` would take precedence over `link_type` for `all`.  We can probably get away without changing `shared` to `link_type`, with the rule:

1. If no tildes are in variant sets to be merged, then the merged set should be the union of the original sets.

2. If there is a tilde somewhere, then treat each variant for which a tilde occurs at least once as an alternative (i.e. as if it were really a multi-valued variant with `multi=False`).  In that case, use precedence rules (as above) to determine whether the `+` or `~` version should be used in the merged list.

3. For multi-valued variants with `multi=True`, merge all the multiple values into one for that variant.

Reply:
> This seems too early to turn the list of variants into string. I think we should ensure they are treated as a list here, then deduplicated, then turned into strings. Does it make sense?

I currently don't do deduplication (see my other comment). If we want the more specific variant to take precedence, your approach probably makes more sense.